pkg/microservice/warpdrive/core/service/taskplugin/deploy.go

```diff
@@ -185,63 +185,133 @@ func (p *DeployTaskPlugin) Run(ctx context.Context, pipelineTask *task.Task, _ *
-						Kind:      setting.Deployment,
-						Container: container.Name,
-						Origin:    container.Image,
-						Name:      deploy.Name,
-					})
-					replaced = true
 				}
 			}
-		}
-
-		for _, sts := range statefulSets {
-			for _, container := range sts.Spec.Template.Spec.Containers {
-				if container.Name == p.Task.ContainerName {
-					err = updater.UpdateStatefulSetImage(sts.Namespace, sts.Name, p.Task.ContainerName, p.Task.Image, p.kubeClient)
-					if err != nil {
-						err = errors.WithMessagef(
-							err,
-							"failed to update container image in %s/statefulsets/%s/%s",
-							p.Task.Namespace, sts.Name, container.Name)
-						return
+		Loop:
+			for _, sts := range statefulSets {
+				for _, container := range sts.Spec.Template.Spec.Containers {
+					if container.Name == p.Task.ContainerName {
+						err = updater.UpdateStatefulSetImage(sts.Namespace, sts.Name, p.Task.ContainerName, p.Task.Image, p.kubeClient)
+						if err != nil {
+							err = errors.WithMessagef(
+								err,
+								"failed to update container image in %s/statefulsets/%s/%s",
+								p.Task.Namespace, sts.Name, container.Name)
+							return
+						}
+						p.Task.ReplaceResources = append(p.Task.ReplaceResources, task.Resource{
+							Kind:      setting.StatefulSet,
+							Container: container.Name,
+							Origin:    container.Image,
+							Name:      sts.Name,
+						})
+						replaced = true
+						break Loop
+					}
+				}
+			}
+		} else {
+			switch serviceInfo.WorkloadType {
+			case setting.StatefulSet:
+				var statefulSet *appsv1.StatefulSet
+				statefulSet, _, err = getter.GetStatefulSet(p.Task.Namespace, p.Task.ServiceName, p.kubeClient)
+				if err != nil {
+					return
+				}
+				for _, container := range statefulSet.Spec.Template.Spec.Containers {
+					if container.Name == p.Task.ContainerName {
+						err = updater.UpdateStatefulSetImage(statefulSet.Namespace, statefulSet.Name, p.Task.ContainerName, p.Task.Image, p.kubeClient)
+						if err != nil {
+							err = errors.WithMessagef(
+								err,
+								"failed to update container image in %s/statefulsets/%s/%s",
+								p.Task.Namespace, statefulSet.Name, container.Name)
+							return
+						}
+						p.Task.ReplaceResources = append(p.Task.ReplaceResources, task.Resource{
+							Kind:      setting.StatefulSet,
+							Container: container.Name,
+							Origin:    container.Image,
+							Name:      statefulSet.Name,
+						})
+						replaced = true
+						break
+					}
+				}
+			case setting.Deployment:
+				var deployment *appsv1.Deployment
+				deployment, _, err = getter.GetDeployment(p.Task.Namespace, p.Task.ServiceName, p.kubeClient)
+				if err != nil {
+					return
+				}
+				for _, container := range deployment.Spec.Template.Spec.Containers {
+					if container.Name == p.Task.ContainerName {
+						err = updater.UpdateDeploymentImage(deployment.Namespace, deployment.Name, p.Task.ContainerName, p.Task.Image, p.kubeClient)
```

Comment:
you can update the image directly, why you need to get the deploy first, and then loop for all containers?

Reply:
fixed